flax/experimental/nnx/nnx/rnglib.py

```diff
@@ -110,44 +120,48 @@ def __init__(
       else:
         rngs['default'] = default
 
-    self._rngs = {
-      name: RngStream(
+    for name, value in rngs.items():
+      stream = RngStream(
+        tag=name,
         key=jax.random.key(value) if isinstance(value, int) else value,
         count=jnp.array(0, dtype=jnp.uint32),
       )
-      for name, value in rngs.items()
-    }
+      setattr(self, name, stream)
 
-  def _make_rng(self, name: str, error_type: Exception) -> jax.Array:
-    if name not in self._rngs:
-      if 'default' not in self._rngs:
+  def _get_stream(self, name: str, error_type: Exception) -> RngStream:
+    rngs_vars = vars(self)
+    if name not in rngs_vars:
+      if 'default' not in rngs_vars:
         raise error_type(f"No RNG named {name!r} or 'default' found in Rngs.")
-      stream = self._rngs['default']
+      stream = rngs_vars['default']
     else:
-      stream = self._rngs[name]
+      stream = rngs_vars[name]
 
-    return stream.make_rng()
+    return stream
 
-  def __getitem__(self, name: str) -> tp.Callable[[], jax.Array]:
-    return lambda: self._make_rng(name, KeyError)
+  def __getitem__(self, name: str):
+    return self._get_stream(name, KeyError)
 
-  def __getattr__(self, name: str) -> tp.Callable[[], jax.Array]:
-    return lambda: self._make_rng(name, AttributeError)
+  def __getattr__(self, name: str):
+    return self._get_stream(name, AttributeError)
 
   def __call__(self):
     return self.default()
 
   def __iter__(self) -> tp.Iterator[str]:
-    return iter(self._rngs)
+    for name in vars(self):
+      if name != '_graph_node__state':
+        yield name
 
   def __len__(self) -> int:
-    return len(self._rngs)
+    return len(vars(self)) - 1
 
   def __contains__(self, name: tp.Any) -> bool:
-    return name in self._rngs
+    return name in vars(self)
 
   def replace(self, **kwargs: tp.Union[int, jax.Array, RngStream]) -> 'Rngs':
-    rngs: dict[str, tp.Any] = self._rngs.copy()
+    rngs: dict[str, tp.Any] = vars(self).copy()
+    del rngs['_graph_node__state']
```

Comment:
It's unfortunate that the implementation detail of GraphNode leaks into the subclasses. Can we do better?